File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/network/CollectorCookieJar.java
package com.snowplowanalytics.snowplow.network;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;

import org.json.JSONException;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

import okhttp3.Cookie;
import okhttp3.HttpUrl;

import static com.snowplowanalytics.snowplow.internal.constants.TrackerConstants.COOKIE_PERSISTANCE;

public class CollectorCookieJar implements okhttp3.CookieJar {
    private Set<CollectorCookie> cookies;
    private final SharedPreferences sharedPreferences;

    public CollectorCookieJar(@NonNull Context context) {
        cookies = Collections.newSetFromMap(new ConcurrentHashMap<CollectorCookie, Boolean>());
        sharedPreferences = context.getSharedPreferences(COOKIE_PERSISTANCE, Context.MODE_PRIVATE);

        loadFromSharedPreferences();
    }

    @NonNull
    @Override
    public List<Cookie> loadForRequest(@NonNull HttpUrl url) {
        List<CollectorCookie> cookiesToRemove = new ArrayList<>();
        List<Cookie> validCookies = new ArrayList<>();

        for (CollectorCookie currentCookie : cookies) {
            if (currentCookie.isExpired()) {
                cookiesToRemove.add(currentCookie);
            } else if (currentCookie.getCookie().matches(url)) {
                validCookies.add(currentCookie.getCookie());
            }
        }

        if (!cookiesToRemove.isEmpty()) {
            removeAll(cookiesToRemove);
        }

        return validCookies;
    }

    @Override
    public void saveFromResponse(@NonNull HttpUrl httpUrl, @NonNull List<Cookie> cookies) {
        saveAll(cookies);
    }

    public void clear() {
        sharedPreferences.edit().clear().apply();
        cookies.clear();
    }

    private void loadFromSharedPreferences() {
        List<String> cookiesToRemove = new ArrayList<>();
        for (Map.Entry<String, ?> entry : sharedPreferences.getAll().entrySet()) {
            String serializedCookie = (String) entry.getValue();

            if (serializedCookie == null) {
                continue;
            }

            try {
                CollectorCookie cookie = new CollectorCookie(serializedCookie);
                cookies.add(cookie);
            } catch (JSONException ignored) {
                cookiesToRemove.add(entry.getKey());
            }
        }

        if (!cookiesToRemove.isEmpty()) {
            SharedPreferences.Editor editor = sharedPreferences.edit();
            for (String cookie : cookiesToRemove) {
                editor.remove(cookie);
            }
            editor.apply();
        }
    }

    private void saveAll(Collection<Cookie> newCookies) {
        SharedPreferences.Editor editor = sharedPreferences.edit();

        for (CollectorCookie cookie : CollectorCookie.decorateAll(newCookies)) {
            cookies.remove(cookie);
            cookies.add(cookie);

            editor.putString(cookie.getCookieKey(), cookie.serialize());
        }

        editor.apply();
    }

    private void removeAll(Collection<CollectorCookie> cookiesToRemove) {
        SharedPreferences.Editor editor = sharedPreferences.edit();

        for (CollectorCookie cookie : cookiesToRemove) {
            cookies.remove(cookie);
            editor.remove(cookie.getCookieKey());
        }

        editor.apply();
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/network/CollectorCookieJar.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.network

import android.content.Context
import android.content.SharedPreferences
import com.snowplowanalytics.core.constants.TrackerConstants
import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.HttpUrl
import org.json.JSONException
import java.util.*
import java.util.concurrent.ConcurrentHashMap

class CollectorCookieJar(context: Context) : CookieJar {
    private val cookies: MutableSet<CollectorCookie> = Collections.newSetFromMap(ConcurrentHashMap())
    private val sharedPreferences: SharedPreferences

    init {
        sharedPreferences =
            context.getSharedPreferences(TrackerConstants.COOKIE_PERSISTANCE, Context.MODE_PRIVATE)
        loadFromSharedPreferences()
    }

    override fun loadForRequest(url: HttpUrl): List<Cookie> {
        val cookiesToRemove: MutableList<CollectorCookie> = ArrayList()
        val validCookies: MutableList<Cookie> = ArrayList()
        
        for (currentCookie in cookies) {
            if (currentCookie.isExpired) {
                cookiesToRemove.add(currentCookie)
            } else if (currentCookie.cookie.matches(url)) {
                validCookies.add(currentCookie.cookie)
            }
        }
        if (cookiesToRemove.isNotEmpty()) {
            removeAll(cookiesToRemove)
        }
        return validCookies
    }

    override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
        saveAll(cookies)
    }

    fun clear() {
        sharedPreferences.edit().clear().apply()
        cookies.clear()
    }

    private fun loadFromSharedPreferences() {
        val cookiesToRemove: MutableList<String> = ArrayList()
        for ((key, value) in sharedPreferences.all) {
            val serializedCookie = value as? String? ?: continue
            try {
                val cookie = CollectorCookie(serializedCookie)
                cookies.add(cookie)
            } catch (ignored: JSONException) {
                cookiesToRemove.add(key)
            }
        }
        
        if (cookiesToRemove.isNotEmpty()) {
            val editor = sharedPreferences.edit()
            for (cookie in cookiesToRemove) {
                editor.remove(cookie)
            }
            editor.apply()
        }
    }

    private fun saveAll(newCookies: Collection<Cookie>) {
        val editor = sharedPreferences.edit()
        for (cookie in CollectorCookie.decorateAll(newCookies)) {
            cookies.remove(cookie)
            cookies.add(cookie)
            editor.putString(cookie.cookieKey, cookie.serialize())
        }
        editor.apply()
    }

    private fun removeAll(cookiesToRemove: Collection<CollectorCookie>) {
        val editor = sharedPreferences.edit()
        for (cookie in cookiesToRemove) {
            cookies.remove(cookie)
            editor.remove(cookie.cookieKey)
        }
        editor.apply()
    }
}
